util/network-devp2p/src/discovery.rs

```diff
@@ -601,51 +664,56 @@ impl<'a> Discovery<'a> {
 	}
 
 	fn check_expired(&mut self, time: Instant) {
-		while let Some((node_id, sent_at)) = self.expiring_pings.pop_front() {
-			if time.duration_since(sent_at) <= PING_TIMEOUT {
-				self.expiring_pings.push_front((node_id, sent_at));
-				break;
+		let mut nodes_to_expire = Vec::new();
+		self.in_flight_pings.retain(|node_id, ping_request| {
+			if time.duration_since(ping_request.sent_at) > PING_TIMEOUT {
+				debug!(target: "discovery", "Removing expired PING request for node_id=0x{:x}", node_id);
+				nodes_to_expire.push(*node_id);
+				false
+			} else {
+				true
 			}
-			self.expire_in_flight_request(node_id, sent_at);
-		}
-		while let Some((node_id, sent_at)) = self.expiring_finds.pop_front() {
-			if time.duration_since(sent_at) <= FIND_NODE_TIMEOUT {
-				self.expiring_finds.push_front((node_id, sent_at));
-				break;
+		});
+		self.in_flight_find_nodes.retain(|node_id, find_node_request| {
+			if time.duration_since(find_node_request.sent_at) > FIND_NODE_TIMEOUT {
+				if !find_node_request.answered {
```

Comment:
The rationale for requiring exactly BUCKET_SIZE results is discussed [here](https://github.com/paritytech/parity-ethereum/pull/8771#discussion_r201172193) and matches behavior in Geth. Can you explain why you prefer not to expire if there is an insufficient number of responses?

Reply:
OK, that makes sense.

[This](https://github.com/ethereum/go-ethereum/blob/c73b654fd172464f598c239b4e7c4bc07a641e3b/p2p/discover/udp.go#L329) [code](https://github.com/ethereum/go-ethereum/blob/c73b654fd172464f598c239b4e7c4bc07a641e3b/p2p/discover/udp.go#L438) in Geth is where it seems to me it would timeout FIND_NODE requests with insufficient results.

Reply:
So I think that the first bit is checking that the result is actually < 16 (it errors if `nreceived >= bucketSize`) ; and the second I'm not 100% sure but I think it's a general timeout machanism, not related to `FindNode`

Reply:
Where do you see that it errors? In my reading of it, [this code](https://github.com/ethereum/go-ethereum/blob/c73b654fd172464f598c239b4e7c4bc07a641e3b/p2p/discover/udp.go#L422), marks the request as complete once `nreceived >= bucketSize` and does nothing (waiting for more responses) if it's not.

Reply:
Hm I think you are right, sorry I misread the code... Even so, I think the mechanism in this PR is better.